# IDS Assignment Part 2 - <font color="red"><h7>Deadline: 23/01/2023 23:59</h7></font>
This is the second part of the assignment in IDS 2022/2023. 
Please use this Jupyter notebook to work on the questions posed in the assignment. When you are done, upload the notebook in Moodle at the designated activity. In addition to the _Jupyter notebook_, please submit _one zip-file_ containing your screenshots for Question 7. 

Give your commented Python code and answers in the corresponding provided cells. Make sure to answer all questions in a clear and explicit manner and discuss your outputs. _Please do not change the general structure of this notebook_. You can, however, add additional markdown or code cells if necessary. <b>Please DO NOT CLEAR THE OUTPUT of the notebook you are submitting! </b>

<font color="red"> *Please make sure to include the names and matriculation numbers of all group members in the slot provided below.* </font> If a name or a student id is missing, the student will not receive any points.

Hint 1: While working on the assignment, you will get a better understanding of the dataset. Feel free to generate additional results and visualizations to support your answers. For example, this might be useful regarding data modification, data simplification, or output interpretation. <font color="red">Ensure that all your claims are supported.</font>

Hint 2: <font color="red">Plan your time wisely. </font> A few parts of this assignment may take some time to run. It might be necessary to consider time management when you plan your group work. Also, do not attempt to upload your assignment at the last minute before the deadline. This often does not work, and you will miss the deadline. Late submissions will not be considered.

Hint 3: RWTHmoodle allows multiple submissions, with every new submission overwriting the previous one. <b>Partial submissions are therefore possible and encouraged. </b> This might be helpful in case of technical issues with RWTHMoodle, which may occur close to the deadline.

<font color="red"><b>Student Names and IDs:
    
    1. 
    
    2. 
    
    3.

## Question 1: Preprocessing (20 points)

In [1]:
import pandas as pd
import numpy as np 

### Sklearn
from sklearn.preprocessing import MinMaxScaler, StandardScaler
from sklearn.impute import KNNImputer

from sklearn.preprocessing import OneHotEncoder

import plotly.express as px

In this question, we consider a dataset documenting the Ski Resorts in Europe (**ski_resorts.csv**).
Each row contains some information about the Ski resort.
You can find a short description for each column:

| Column | Description |
| ------ | ----------- |
| Resort | The name of the ski & snowboard resort. |
| Country | The country in which the resort is located. |
| HighestPoint | The highest mountain point at the ski resort.   |
| LowestPoint | The lowest possible point to ski at the ski resort.  |
| DayPassPriceAdult | The price shows what it costs for 1 adult for 1 day in the main season in Euro. |
| BeginnerSlope | The total amount of “beginner” slopes in kilometer at the resort. “Beginner slopes” contains “children”, “blue” and, “green” slopes. |
| IntermediateSlope | The total amount of “intermediate” slopes in kilometer at the resort. “Intermediate slopes” contains “red” slopes. |
| DifficultSlope | The total amount of “difficult” slopes in kilometer at the resort. “Difficult slopes” contains “black”, “advanced”, and ”expert” slopes. |
| TotalSlope | The sum of “beginner slopes” + “intermediate slopes” + “difficult slopes” |
| Snowparks | Does the resort have one or more snowparks, or not? |
| NightSki | Does the resort offer skiing on illuminated slopes? |
| SurfaceLifts | The amount of lifts in this category: T-bar, Sunkidslift, Rope lifts, and people mower. |
| ChairLifts | The total amount of chairlifts. |
| GondolaLifts | The amount of lifts in this category: Gondola, Train lifts, Funicular, Combined gondola and chairlifts, Helicopter lifts, Snowcats, and Aerial tramways. |
| TotalLifts | The sum of “surface lifts etc” + “gondola etc” + “chairlifts etc.” |
| LiftCapacity | How many passengers can the lift system at the ski resort mowe in one hour? |
| SnowCannons  |The total amount of snow cannons at the ski resort.|

### a) Loading the Data and Initial Quality Investigation (2.5pts)

#### **a(i)** 
Load the dataset into a dataframe `df`. <font color='red'>Use the first column as index for your dataframe</font>. Ensure that the index is valid, that is, it should not contain any duplicate entries. 



**In the subsequent questions, only modify the dataframe `df` if explicitly requested. However, you can always create working copies.**

In [2]:
# your code
df = pd.read_csv('ski_resorts.csv', index_col=0)     # load data set into df with first col as index
# type(df)

df.index.duplicated().sum()                          # no duplicates in index -> index is valid


0

#### **a(ii)** 
Show the data types of the dataframe columns as well as the first 5 rows. On the first sight, are there any data type problems (e.g., numerical columns having a non-numerical data type)?



In [3]:
# your code

for col in df.columns:
    print(col, type(df[col]))
    for i in range(1,6):
        print(type(df[col][i]))

Resort <class 'pandas.core.series.Series'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
Country <class 'pandas.core.series.Series'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
HighestPoint <class 'pandas.core.series.Series'>
<class 'numpy.float64'>
<class 'numpy.float64'>
<class 'numpy.float64'>
<class 'numpy.float64'>
<class 'numpy.float64'>
LowestPoint <class 'pandas.core.series.Series'>
<class 'numpy.float64'>
<class 'numpy.float64'>
<class 'numpy.float64'>
<class 'numpy.float64'>
<class 'numpy.float64'>
DayPassPriceAdult <class 'pandas.core.series.Series'>
<class 'numpy.float64'>
<class 'numpy.float64'>
<class 'numpy.float64'>
<class 'numpy.float64'>
<class 'numpy.float64'>
BeginnerSlope <class 'pandas.core.series.Series'>
<class 'numpy.float64'>
<class 'numpy.float64'>
<class 'numpy.float64'>
<class 'numpy.float64'>
<class 'numpy.float64'>
IntermediateSlope <class 'pandas.core.series.Series'>
<class 'numpy.int64'>
<class 'numpy.in

**Your Answer:** There does not seem to be any problem with inconsintencies, however, it is stricing, that there are no bools, although one would expect that from the above given desciption, i.e. 'Nightski' or 'SnowPark'.

#### **a(iii)** 
To improve performance and memory usage (in particular for large datasets) it is important to use **categorical** columns whenever suitable.
Are there any categorical column candidates? Explain your answer. \
Afterward, convert the column(s) in `df` into categorical column(s).

In [4]:
df.head()


Resort          Country  \
1                              Alpendorf (Ski amedé)          Austria   
2  Soldeu-Pas de la Casa/​Grau Roig/​El Tarter/​C...          Andorra   
3                              Oberau (Wildschönau)          Austria   
4                                     Dachstein West          Austria   
5                                        Rosa Khutor  Southern Russia   

   HighestPoint  LowestPoint  DayPassPriceAdult  BeginnerSlope  \
1        1980.0        740.0               52.0           30.0   
2        2640.0       1710.0               47.0          100.0   
3        1130.0        900.0               30.0            1.0   
4        1620.0        780.0               42.0           15.0   
5        2320.0        940.0               22.0           30.0   

   IntermediateSlope  DifficultSlope  TotalSlope Snowparks NightSki  \
1                 81             4.0         115       Yes       No   
2                 77            33.0         210       Yes      Yes   
3                  0             1.0           2        No       No   
4                 33             3.0          51       Yes      Yes   
5                 26            21.0          77       Yes       No   

   SurfaceLifts  ChairLifts  GondolaLifts  TotalLifts  LiftCapacity  \
1          22.0          16          11.0          49       75398.0   
2          37.0          28           7.0          72       99017.0   
3           2.0           0           0.0           2        1932.0   
4          25.0           8           3.0          36       32938.0   
5           6.0          11          10.0          27       49228.0   

   SnowCannons  
1        600.0  
2       1032.0  
3          0.0  
4        163.0  
5        450.0

In [5]:
# your code
# df.head()
for col in df.columns:
    print(col, len(df[col].unique()))       # shows how many unique entries there are for each column of df

print('total number of rows: ', df.shape[0])

Resort 376
Country 27
HighestPoint 277
LowestPoint 239
DayPassPriceAdult 53
BeginnerSlope 83
IntermediateSlope 86
DifficultSlope 45
TotalSlope 129
Snowparks 2
NightSki 2
SurfaceLifts 48
ChairLifts 37
GondolaLifts 24
TotalLifts 67
LiftCapacity 329
SnowCannons 153
total number of rows:  376


**Your Answer:** The overview above shows that there are only few candidates for being converted into categorical data, because the variety of their entries is too high compared to there being 376 rows in total.
Let's have a look at the distributions via histograms:

In [6]:
num_cols = ['HighestPoint','LowestPoint','DayPassPriceAdult','BeginnerSlope','IntermediateSlope','DifficultSlope','TotalSlope','SurfaceLifts','ChairLifts','GondolaLifts','TotalLifts','LiftCapacity','SnowCannons']
for col in num_cols:
    fig = px.histogram(df, x=col,nbins=100)
    fig.show()

We would categrorize: 'DayPassPriceAdult', 'HighestPoint', 'LowestPoint', 'LiftCapacity' and 'SnowCannons' accordingly to their distribution shown by the histogram above. \
The other numerical variables are left as numerical because they depend on each other in a linear way (see description of variables above). This means, that mathematical operations as + and - should remain on these variables, so we need them to stay numerical.\
This way, we remain with the least numerical variables possible while retaining the properties needed on the others.

In [7]:
df_cat = df.copy()      # make copy of df as we change the entries of the variables

# DayPassPriceAdult
bins_DPPA = [0,0.01, 29.999, 54.999, 1000]
group_names_DPPA = ['free','cheap','medium','expensive']
df_cat['DayPassPriceAdult'] = pd.cut(df['DayPassPriceAdult'], bins_DPPA, labels=group_names_DPPA)
#print (df_cat['DayPassPriceAdult'])

# HighestPoint
bins_HP = [0,500, 1000, 1500, 2000, 2500, 3000, 3500, 4000]
group_names_HP = ['[0,500]','(500,1000]','(1000,1500]','(1500,2000]','(2000,2500]','(2500,3000]','(3000,3500]','(3500,4000]']
df_cat['HighestPoint'] = pd.cut(df['HighestPoint'], bins_HP, labels=group_names_HP)
# print(df_cat['HighestPoint'])

# LowestPoint
bins_LP = [0,500, 1000, 1500, 2000, 2500]
group_names_LP = ['[0,500]','(500,1000]','(1000,1500]','(1500,2000]','(2000,2500]']
df_cat['LowestPoint'] = pd.cut(df['LowestPoint'], bins_LP, labels=group_names_LP)
# print(df_cat['LowestPoint'])

# LiftCapacity
bins_LC = [0,15000, 50000, 100000, 150000, 300000]
group_names_LC = ['[0,15000]','(1500,50000]','(50000,100000]','(100000,150000]','(150000,300000]']
df_cat['LiftCapacity'] = pd.cut(df['LiftCapacity'], bins_LC, labels=group_names_LC)
# print(df_cat['LiftCapacity'])

# SnowCannons
bins_SC = [0,49, 399, 999, 1699, 2400]
group_names_SC = ['[0,49]','[50,399]','[400,999]','[1000,1699]','[1700,2400]']
df_cat['SnowCannons'] = pd.cut(df['SnowCannons'], bins_SC, labels=group_names_SC)
# print(df_cat['SnowCannons'])

In [8]:
df_cat.dtypes

Resort                 object
Country                object
HighestPoint         category
LowestPoint          category
DayPassPriceAdult    category
BeginnerSlope         float64
IntermediateSlope       int64
DifficultSlope        float64
TotalSlope              int64
Snowparks              object
NightSki               object
SurfaceLifts          float64
ChairLifts              int64
GondolaLifts          float64
TotalLifts              int64
LiftCapacity         category
SnowCannons          category
dtype: object

### b) Handling Missing Values & Encoding (17.5pts)

In the following task, you can assume that every NAN entry in the dataframe is actually a missing value. This can partially be justified by the fact that pandas did not have problems inferring the "proper" datatypes (e.g., a string indicating a missing number in a number column would result in pandas parsing an object column) and your subsequent check of the data types. Therefore, you can use `df.isna()` as a proxy indicator for missing values.

#### **b(i)** 
Simply discarding missing entries is usually not a good idea. Therefore, you should first analyze the number of missing values and check for patterns of missing values. 

To this end, compute the following statistics on missing values:
1. How many entries does the dataframe have? (To relate this to the number of entries missing)
2. How many missing values do we have? What is the ratio i.e., "number of missing values"/"number of entries of df"?
3. How many rows have at least a single missing value?
4. Count the number of missing values per column.
5. Count the number of missing values per row and aggregate them - i.e., show the number of rows that suffer from x missing values.
6. What do you observe? Are there any rows containing missing values for the same set of columns? Can you identify potential patterns?


In [9]:
# your code for 1. How many entries does the dataframe have? (To relate this to the number of entries missing)
print('overall total number of entries in df: ',(df.shape[0]*df.shape[1]))
# print('overall total number of missings in df: ',df.isna().sum().sum())

overall total number of entries in df:  6392


In [10]:
# your code for 2. How many missing values do we have? What is the ratio i.e., "number of missing values"/"number of entries of df"?
print('overall total number of missings in df: ',df.isna().sum().sum())
print('overall missing ratio in df: ',(df.isna().sum().sum())/(df.shape[0]*df.shape[1]))
# print('missing ratios by coloumn:')
# (df.isna().sum())/df.shape[0]

overall total number of missings in df:  87
overall missing ratio in df:  0.013610763454317897


In [11]:
# print('column names with at least one missing value:')
# count_int=0
# for col in df.columns:
#     if df[col].isna().sum()>0:
#         count_int+=1
#         print(col)
# print('Number of columns with at least a single missing value: ',count_int)

In [12]:
# your code for 3. How many rows have at least a single missing value?
# print('row names with at least one missing value:')
count_int=0
for row in range(0,df.shape[0]):
    if df.iloc[row].isna().sum()>0:
        count_int+=1
        # print(row)
print('Number of rows with at least a single missing value: ',count_int)

Number of rows with at least a single missing value:  45


In [13]:
# your code for 4. Count the number of missing values per column.
print('total count of missings by coloumn:')
df.isna().sum()

total count of missings by coloumn:


Resort                0
Country               0
HighestPoint         13
LowestPoint           3
DayPassPriceAdult    17
BeginnerSlope         8
IntermediateSlope     0
DifficultSlope        8
TotalSlope            0
Snowparks             0
NightSki              0
SurfaceLifts          6
ChairLifts            0
GondolaLifts          7
TotalLifts            0
LiftCapacity         17
SnowCannons           8
dtype: int64

In [14]:
# your code for 5. Count the number of missing values per row and aggregate them - i.e., show the number of rows that suffer from x missing values.
# df.isna().sum(axis=1)
# max(df.isna().sum(axis=1).tolist())

for i in range(0,max(df.isna().sum(axis=1).tolist())+1):
    print(i,df.isna().sum(axis=1).tolist().count(i))

0 331
1 35
2 2
3 0
4 0
5 0
6 8


In [15]:
# added code for answering 6:

null_df = df[df.isna().any(axis=1)]   # extracting only rows of df with >0 missing values

# print('row names with at least one missing value:')
# count_int=0
# for row in range(0,null_df.shape[0]):
#     if null_df.iloc[row].isna().sum()>0:
#         count_int+=1
#         # print(col)
#         res_int = [i for i, val in enumerate(null_df.iloc[row].isna().tolist()) if val]
#         res_str = null_df.columns[res_int]
#         print(row, str(res_str))
# print('Number of rows with at least a single missing value: ',count_int)


print('column names with at least one missing value:')
count_int=0
for col in null_df.columns:
    if null_df[col].isna().sum()>0:
        count_int+=1
        # print(col)
        res = [i for i, val in enumerate(null_df[col].isna().tolist()) if val]
        print(col, str(res))
print('Number of columns with at least a single missing value: ',count_int)

column names with at least one missing value:
HighestPoint [1, 10, 13, 20, 21, 22, 23, 24, 25, 26, 28, 29, 42]
LowestPoint [33, 34, 41]
DayPassPriceAdult [0, 6, 8, 14, 15, 18, 20, 21, 22, 23, 24, 25, 26, 35, 36, 42, 43]
BeginnerSlope [20, 21, 22, 23, 24, 25, 26, 42]
DifficultSlope [20, 21, 22, 23, 24, 25, 26, 42]
SurfaceLifts [2, 3, 9, 18, 37, 44]
GondolaLifts [7, 12, 16, 17, 19, 32, 39]
LiftCapacity [4, 5, 9, 11, 20, 21, 22, 23, 24, 25, 26, 27, 30, 31, 38, 40, 42]
SnowCannons [20, 21, 22, 23, 24, 25, 26, 42]
Number of columns with at least a single missing value:  9


**Your answer:** *(for 6. What do you observe? Are there any rows containing missing values for the same set of columns?)*
rows 20,21,22,23,24,25,26 and 42 all (only) miss values for 'HighestPoint','DayPassPriceAdult','BeginnerSlope','DifficultSlope','LiftCapacity' and 'SnowCannons'. These are also to rows missing the most values (as can be seen in the solution for question 5).
Row 9 and 18 are the other two having more then 1 missing entry, however they do not form a pattern.
The rest of the rows only has one missing entry and can therefore not form any pattern.

#### **b(ii)**
For the next step:

1. Transform the categorical column(s) you identified in a(iii) into one-hot encoding format. 
2. Transform the columns "Snowparks" and "NightSki" in `df` into boolean data type, where "Yes" should be `True` and "No" should be `False`

In the end, the original categorical column(s) should still be there. Additionally, there should be x number (x is the number of unique values) of one-hot encoding columns for each categorical column. Use the following naming convention for the new columns "{name of the categorical column}_{unique value for that column}" Also, make sure the columns "Snowparks" and "NightSki" are boolean type in the end.
Lastly, print the top five rows of the resulting dataframe.

*Hint: You can use the pd.get_dummies() function from pandas for the first transformation*

In [16]:
# We will use a copy of the df in the following because we will transform the coloumns
df_imp = df_cat.copy()

In [17]:
# your code for 1. Transform the categorical column(s) you identified in a(iii) into one-hot encoding format. 

# list of categorized columns
list_cat = ['DayPassPriceAdult','HighestPoint','LowestPoint','LiftCapacity','SnowCannons']

# only wand to OHE the former categorized columns
X = df_imp[list_cat]

# create encoder and fit and transform on data X, safe all to new encoder_df while keeping old index and clever names :)
enc = OneHotEncoder(handle_unknown='ignore')    # When an unknown category is encountered during transform, the resulting one-hot encoded columns for this feature will be all zeros.
encoded = enc.fit_transform(X).toarray()
encoder_df = pd.DataFrame(encoded,columns=enc.get_feature_names_out(list_cat),index=df_cat.index)
# print(encoder_df)

# merge one-hot encoded columns back with df_cat
df_imp = df_cat.join(encoder_df)

# drop categorized columns NO BECAUSE TASK SAYS TO LEAVE THEM THERE
# df_imp.drop(list_cat, axis=1, inplace=True)

# print(df_imp.head())


In [18]:
# your code for 2. Transform the columns "Snowparks" and "NightSki" in `df` into boolean data type, where "Yes" should be `True` and "No" should be `False`

df_imp.Snowparks = (pd.Series(np.where(df_imp.Snowparks.values == 'Yes', bool(['True']), bool(['False'])),
          df_imp.index))
          
df_imp.NightSki = (pd.Series(np.where(df_imp.NightSki.values == 'Yes', bool(['True']), bool(['False'])),
          df_imp.index))

In [19]:
# use df.dtypes to check if you correctly transform the data
df_imp.dtypes

Resort                            object
Country                           object
HighestPoint                    category
LowestPoint                     category
DayPassPriceAdult               category
BeginnerSlope                    float64
IntermediateSlope                  int64
DifficultSlope                   float64
TotalSlope                         int64
Snowparks                           bool
NightSki                            bool
SurfaceLifts                     float64
ChairLifts                         int64
GondolaLifts                     float64
TotalLifts                         int64
LiftCapacity                    category
SnowCannons                     category
DayPassPriceAdult_cheap          float64
DayPassPriceAdult_expensive      float64
DayPassPriceAdult_medium         float64
DayPassPriceAdult_nan            float64
HighestPoint_(1000,1500]         float64
HighestPoint_(1500,2000]         float64
HighestPoint_(2000,2500]         float64
HighestPoint_(25

In [20]:
#(df_imp['SnowCannons_[50,399]'])

#### **b(iii)** 
The previous analysis in b(i) showed that there are missing values in the 'SurfaceLifts' and 'GondolaLifts' columns.\
How would you impute these values? \
Motivate your approach and apply it to `df`.

*Hint: Remember the semantics of the columns. Also, carefully assert your assumptions.*


**Your Answer:** From the description of the variables we know, that 'SurfaceLifts' + 'GondolaLifts' + 'ChairLifts' = 'TotalLifts'. Therefore, for every row with only one of the four entries missing, we will calculate it via the other two and the given formula. \
From task 1b (1) Question 6, we already know, that there are no missings in 'ChairLifts' and 'TotalLifts'. Additionally this task showed explicitly, that the missing rows of 'SurfaceLifts' and 'GondolaLifts' are disjoint. \
Hence, we can calculate the replacing value as follows:

In [21]:
# your code

if df_imp['SurfaceLifts'].isna().sum()>0:
        # count_int+=1
        # print(col)
        res_SL = [i for i, val in enumerate(df_imp['SurfaceLifts'].isna().tolist()) if val]
        print('SurfaceLifts', str(res_SL))

# SurfaceLifts = TotalLifts - (GondolaLifts + ChairLifts)
for row in res_SL:
        df_imp['SurfaceLifts'][row] = df_imp['TotalLifts'][row]-(df_imp['GondolaLifts'][row]+df_imp['ChairLifts'][row])
        print(df_imp['SurfaceLifts'][row])

# impute GondolaLifts
if df_imp['GondolaLifts'].isna().sum()>0:
        # count_int+=1
        # print(col)
        res_GL = [i for i, val in enumerate(df_imp['GondolaLifts'].isna().tolist()) if val]
        print('GondolaLifts', str(res_GL))

# GondolaLifts = TotalLifts - (SurfaceLifts + ChairLifts)
for row in res_GL:
        df_imp['GondolaLifts'][row] = df_imp['TotalLifts'][row]-(df_imp['SurfaceLifts'][row]+df_imp['ChairLifts'][row])
        print(df_imp['GondolaLifts'][row])




SurfaceLifts [22, 49, 118, 203, 274, 371]
32.0
8.0
9.0
2.0
18.0
62.0
GondolaLifts [104, 125, 188, 200, 207, 249, 297]
10.0
0.0
0.0
0.0
0.0
0.0
4.0


C:\Users\kfeld\AppData\Local\Temp/ipykernel_772/3519041251.py:11: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy

C:\Users\kfeld\AppData\Local\Temp/ipykernel_772/3519041251.py:23: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



#### **b(iv)**
Impute the rest of the missing values using the knn-imputation method. To this end, apply the following steps:
1. Create a working copy `df_tmp` of your updated `df`.
2. For simplicity, drop the non-numerical columns (i.e., not of types integer or floats), which also inlcude the one-hot encoded and the boolean columns* you created earlier.
3. Normalize the data in `df_tmp` (e.g., Standard score normalization). If the features have very different scales, knn can become very biased.
4. Impute the missing values considering six neighbors.
5. Invert the transformation applied upfront to enable more meaningful and intuitive visualizations.
6. Append the columns you dropped in step 2.
 
In the end, `df` should not contain missing values and have all the columns.

\*Note that by dropping the columns we lose the information of countries and the two boolean attributes ("Snowparks" and "NightSki") when imputing the missing values, which might be crucial for inferencing values such as the price for a ski pass. In practice, one should try to find if there are correlations before deciding whether to drop the columns or not.
We drop the columns here to make the following steps easier because we only have to deal with numerical columns.

*Hint: Be careful with the indices of your dataframes.*


In [22]:
# your code for step 1 & 2
df_tmp = df_imp.copy()._get_numeric_data()
df_tmp.head()

BeginnerSlope  IntermediateSlope  DifficultSlope  TotalSlope  Snowparks  \
1           30.0                 81             4.0         115       True   
2          100.0                 77            33.0         210       True   
3            1.0                  0             1.0           2       True   
4           15.0                 33             3.0          51       True   
5           30.0                 26            21.0          77       True   

   NightSki  SurfaceLifts  ChairLifts  GondolaLifts  TotalLifts  ...  \
1      True          22.0          16          11.0          49  ...   
2      True          37.0          28           7.0          72  ...   
3      True           2.0           0           0.0           2  ...   
4      True          25.0           8           3.0          36  ...   
5      True           6.0          11          10.0          27  ...   

   LiftCapacity_(150000,300000]  LiftCapacity_(50000,100000]  \
1                           0.0                          1.0   
2                           0.0                          1.0   
3                           0.0                          0.0   
4                           0.0                          0.0   
5                           0.0                          0.0   

   LiftCapacity_[0,15000]  LiftCapacity_nan  SnowCannons_[0,49]  \
1                     0.0               0.0                 0.0   
2                     0.0               0.0                 0.0   
3                     1.0               0.0                 0.0   
4                     0.0               0.0                 0.0   
5                     0.0               0.0                 0.0   

   SnowCannons_[1000,1699]  SnowCannons_[1700,2400]  SnowCannons_[400,999]  \
1                      0.0                      0.0                    1.0   
2                      1.0                      0.0                    0.0   
3                      0.0                      0.0                    0.0   
4                      0.0                      0.0                    0.0   
5                      0.0                      0.0                    1.0   

   SnowCannons_[50,399]  SnowCannons_nan  
1                   0.0              0.0  
2                   0.0              0.0  
3                   0.0              1.0  
4                   1.0              0.0  
5                   0.0              0.0  

[5 rows x 41 columns]

In [23]:
# your code for step 3
# we will use Min Max Normailsation:
scaler = MinMaxScaler()
scaled = scaler.fit_transform(df_tmp)
scaled_df = pd.DataFrame(scaled, columns=df_tmp.columns, index=df_tmp.index)
scaled_df.head()

BeginnerSlope  IntermediateSlope  DifficultSlope  TotalSlope  Snowparks  \
1       0.096154           0.338912        0.038835    0.191667        0.0   
2       0.320513           0.322176        0.320388    0.350000        0.0   
3       0.003205           0.000000        0.009709    0.003333        0.0   
4       0.048077           0.138075        0.029126    0.085000        0.0   
5       0.096154           0.108787        0.203883    0.128333        0.0   

   NightSki  SurfaceLifts  ChairLifts  GondolaLifts  TotalLifts  ...  \
1       0.0      0.247191    0.216216         0.275    0.281609  ...   
2       0.0      0.415730    0.378378         0.175    0.413793  ...   
3       0.0      0.022472    0.000000         0.000    0.011494  ...   
4       0.0      0.280899    0.108108         0.075    0.206897  ...   
5       0.0      0.067416    0.148649         0.250    0.155172  ...   

   LiftCapacity_(150000,300000]  LiftCapacity_(50000,100000]  \
1                           0.0                          1.0   
2                           0.0                          1.0   
3                           0.0                          0.0   
4                           0.0                          0.0   
5                           0.0                          0.0   

   LiftCapacity_[0,15000]  LiftCapacity_nan  SnowCannons_[0,49]  \
1                     0.0               0.0                 0.0   
2                     0.0               0.0                 0.0   
3                     1.0               0.0                 0.0   
4                     0.0               0.0                 0.0   
5                     0.0               0.0                 0.0   

   SnowCannons_[1000,1699]  SnowCannons_[1700,2400]  SnowCannons_[400,999]  \
1                      0.0                      0.0                    1.0   
2                      1.0                      0.0                    0.0   
3                      0.0                      0.0                    0.0   
4                      0.0                      0.0                    0.0   
5                      0.0                      0.0                    1.0   

   SnowCannons_[50,399]  SnowCannons_nan  
1                   0.0              0.0  
2                   0.0              0.0  
3                   0.0              1.0  
4                   1.0              0.0  
5                   0.0              0.0  

[5 rows x 41 columns]

In [24]:
# your code for step 4
imputer = KNNImputer(n_neighbors=6)
df_imp_scaled = pd.DataFrame(imputer.fit_transform(scaled_df), columns=scaled_df.columns, index=scaled_df.index)
df_imp_scaled.head()

BeginnerSlope  IntermediateSlope  DifficultSlope  TotalSlope  Snowparks  \
1       0.096154           0.338912        0.038835    0.191667        0.0   
2       0.320513           0.322176        0.320388    0.350000        0.0   
3       0.003205           0.000000        0.009709    0.003333        0.0   
4       0.048077           0.138075        0.029126    0.085000        0.0   
5       0.096154           0.108787        0.203883    0.128333        0.0   

   NightSki  SurfaceLifts  ChairLifts  GondolaLifts  TotalLifts  ...  \
1       0.0      0.247191    0.216216         0.275    0.281609  ...   
2       0.0      0.415730    0.378378         0.175    0.413793  ...   
3       0.0      0.022472    0.000000         0.000    0.011494  ...   
4       0.0      0.280899    0.108108         0.075    0.206897  ...   
5       0.0      0.067416    0.148649         0.250    0.155172  ...   

   LiftCapacity_(150000,300000]  LiftCapacity_(50000,100000]  \
1                           0.0                          1.0   
2                           0.0                          1.0   
3                           0.0                          0.0   
4                           0.0                          0.0   
5                           0.0                          0.0   

   LiftCapacity_[0,15000]  LiftCapacity_nan  SnowCannons_[0,49]  \
1                     0.0               0.0                 0.0   
2                     0.0               0.0                 0.0   
3                     1.0               0.0                 0.0   
4                     0.0               0.0                 0.0   
5                     0.0               0.0                 0.0   

   SnowCannons_[1000,1699]  SnowCannons_[1700,2400]  SnowCannons_[400,999]  \
1                      0.0                      0.0                    1.0   
2                      1.0                      0.0                    0.0   
3                      0.0                      0.0                    0.0   
4                      0.0                      0.0                    0.0   
5                      0.0                      0.0                    1.0   

   SnowCannons_[50,399]  SnowCannons_nan  
1                   0.0              0.0  
2                   0.0              0.0  
3                   0.0              1.0  
4                   1.0              0.0  
5                   0.0              0.0  

[5 rows x 41 columns]

In [25]:
# your code for step 5
df_imputed = pd.DataFrame(scaler.inverse_transform(df_imp_scaled), columns=df_imp_scaled.columns, index=df_imp_scaled.index)
df_imputed.head()

BeginnerSlope  IntermediateSlope  DifficultSlope  TotalSlope  Snowparks  \
1           30.0               81.0             4.0       115.0        1.0   
2          100.0               77.0            33.0       210.0        1.0   
3            1.0                0.0             1.0         2.0        1.0   
4           15.0               33.0             3.0        51.0        1.0   
5           30.0               26.0            21.0        77.0        1.0   

   NightSki  SurfaceLifts  ChairLifts  GondolaLifts  TotalLifts  ...  \
1       1.0          22.0        16.0          11.0        49.0  ...   
2       1.0          37.0        28.0           7.0        72.0  ...   
3       1.0           2.0         0.0           0.0         2.0  ...   
4       1.0          25.0         8.0           3.0        36.0  ...   
5       1.0           6.0        11.0          10.0        27.0  ...   

   LiftCapacity_(150000,300000]  LiftCapacity_(50000,100000]  \
1                           0.0                          1.0   
2                           0.0                          1.0   
3                           0.0                          0.0   
4                           0.0                          0.0   
5                           0.0                          0.0   

   LiftCapacity_[0,15000]  LiftCapacity_nan  SnowCannons_[0,49]  \
1                     0.0               0.0                 0.0   
2                     0.0               0.0                 0.0   
3                     1.0               0.0                 0.0   
4                     0.0               0.0                 0.0   
5                     0.0               0.0                 0.0   

   SnowCannons_[1000,1699]  SnowCannons_[1700,2400]  SnowCannons_[400,999]  \
1                      0.0                      0.0                    1.0   
2                      1.0                      0.0                    0.0   
3                      0.0                      0.0                    0.0   
4                      0.0                      0.0                    0.0   
5                      0.0                      0.0                    1.0   

   SnowCannons_[50,399]  SnowCannons_nan  
1                   0.0              0.0  
2                   0.0              0.0  
3                   0.0              1.0  
4                   1.0              0.0  
5                   0.0              0.0  

[5 rows x 41 columns]

In [26]:
df_imputed.columns.tolist()

['BeginnerSlope',
 'IntermediateSlope',
 'DifficultSlope',
 'TotalSlope',
 'Snowparks',
 'NightSki',
 'SurfaceLifts',
 'ChairLifts',
 'GondolaLifts',
 'TotalLifts',
 'DayPassPriceAdult_cheap',
 'DayPassPriceAdult_expensive',
 'DayPassPriceAdult_medium',
 'DayPassPriceAdult_nan',
 'HighestPoint_(1000,1500]',
 'HighestPoint_(1500,2000]',
 'HighestPoint_(2000,2500]',
 'HighestPoint_(2500,3000]',
 'HighestPoint_(3000,3500]',
 'HighestPoint_(3500,4000]',
 'HighestPoint_(500,1000]',
 'HighestPoint_[0,500]',
 'HighestPoint_nan',
 'LowestPoint_(1000,1500]',
 'LowestPoint_(1500,2000]',
 'LowestPoint_(2000,2500]',
 'LowestPoint_(500,1000]',
 'LowestPoint_[0,500]',
 'LowestPoint_nan',
 'LiftCapacity_(100000,150000]',
 'LiftCapacity_(1500,50000]',
 'LiftCapacity_(150000,300000]',
 'LiftCapacity_(50000,100000]',
 'LiftCapacity_[0,15000]',
 'LiftCapacity_nan',
 'SnowCannons_[0,49]',
 'SnowCannons_[1000,1699]',
 'SnowCannons_[1700,2400]',
 'SnowCannons_[400,999]',
 'SnowCannons_[50,399]',
 'SnowCanno

In [27]:
# your code for step 6
non_num_col = ['Resort', 'Country','Snowparks', 'NightSki','DayPassPriceAdult','HighestPoint','LowestPoint','LiftCapacity','SnowCannons']
df_complete_imp = pd.concat([df_imputed,df_cat[non_num_col]],axis=1)
df_complete_imp.head()

BeginnerSlope  IntermediateSlope  DifficultSlope  TotalSlope  Snowparks  \
1           30.0               81.0             4.0       115.0        1.0   
2          100.0               77.0            33.0       210.0        1.0   
3            1.0                0.0             1.0         2.0        1.0   
4           15.0               33.0             3.0        51.0        1.0   
5           30.0               26.0            21.0        77.0        1.0   

   NightSki  SurfaceLifts  ChairLifts  GondolaLifts  TotalLifts  ...  \
1       1.0          22.0        16.0          11.0        49.0  ...   
2       1.0          37.0        28.0           7.0        72.0  ...   
3       1.0           2.0         0.0           0.0         2.0  ...   
4       1.0          25.0         8.0           3.0        36.0  ...   
5       1.0           6.0        11.0          10.0        27.0  ...   

   SnowCannons_nan                                             Resort  \
1              0.0                              Alpendorf (Ski amedé)   
2              0.0  Soldeu-Pas de la Casa/​Grau Roig/​El Tarter/​C...   
3              1.0                              Oberau (Wildschönau)   
4              0.0                                     Dachstein West   
5              0.0                                        Rosa Khutor   

           Country  Snowparks  NightSki  DayPassPriceAdult  HighestPoint  \
1          Austria        Yes        No             medium   (1500,2000]   
2          Andorra        Yes       Yes             medium   (2500,3000]   
3          Austria         No        No             medium   (1000,1500]   
4          Austria        Yes       Yes             medium   (1500,2000]   
5  Southern Russia        Yes        No              cheap   (2000,2500]   

   LowestPoint    LiftCapacity  SnowCannons  
1   (500,1000]  (50000,100000]    [400,999]  
2  (1500,2000]  (50000,100000]  [1000,1699]  
3   (500,1000]       [0,15000]          NaN  
4   (500,1000]    (1500,50000]     [50,399]  
5   (500,1000]    (1500,50000]    [400,999]  

[5 rows x 50 columns]

In [28]:
df_complete_imp.isna().sum().sum() == 0

False

## Question 2: Visualization (13 points)

In this task, you will analyze the data using different means of visualization.

Start with the following preprocessed and integrated dataframe `df_v`. \
Note that it has a similar structure to the dataframe that you should obtain from the previous task, however, the values have been modified.

In [ ]:
import pandas as pd
import numpy as np
import plotly.express as px
from pandas.plotting import scatter_matrix
import matplotlib.pyplot as plt
import seaborn as sns

In [ ]:
df_v = pd.read_csv("./datasets/ski_resorts_visual.csv", index_col=0)
df_num = df_v.copy()._get_numeric_data()    # df_v with only numeric colomns for a) and b)
# df_num.head()

### **a) (3pts)** 
To start the visual analysis, make a Scatter plot matrix to visually check if there are any correlations between the numerical attributes.

*Hint: You can use the scatter_matrix from pandas.plotting or pairplot from seaborn to make the plot.*


In [ ]:
# your code
sns.pairplot(df_num)      # automatically only uses numerical data columns

### **b) (3pts)**
Another way to detect correlation is to calculate the Pearson correlation coefficient. Calculate the correlation matrix for the numerical data and visualize the matrix using a heatmap. 
Briefly discuss your findings from the heatmap and the scatter plot you created in 2(a).

Make sure to annotate the heatmap with the values of the correlation.

*Hint: You can use the heatmap function from seaborn to make the plot.*

In [ ]:
# your code
corr_matrix = df_num.corr()
# print(corr_matrix)

sns.set(rc={"figure.figsize":(15, 15)})
sns.heatmap(corr_matrix, annot=True, fmt=".2f")#, cmap="crest")

#### Oberservations:

### **c) (4pts)** 
We now want to have an overview of the attribute "TotalSlope" aggregated by different levels of hierarchy (Europe -> Country -> Resort). It seems that a tree map is suitable for this purpose.

Make a tree map where
- the root node represents Europe.
- the child nodes of Europe are countries.
- the child nodes of each country are the ski resorts.
- the size of the rectangles is determined by the attribute "TotalSlope".

Also, use the tree map to find out
1. The sum of TotalSlopes of a country, list the top five countries and the corresponding values.
2. The max value of TotalSlope of the five countries you identified in 1.

*Hint: You can use the treemap function from plotly.express.*

In [ ]:
# All countries are in Europe
# df_v.Country.unique()

In [ ]:
# your code
fig = px.treemap(df_v, path=[px.Constant("Europe"), 'Country', 'Resort'], values='TotalSlope')
fig.update_traces(root_color="lightgrey")
fig.update_layout(margin = dict(t=50, l=25, r=25, b=25))
fig.show()

**Your answer for...** \
*...  1. The sum of TotalSlopes of a country. List the top five countries and the corresponding values:* 
Top five countries in descending order with corresponting values of total slopes in brackets: France (12546), Austria (7005), Switzerland (5710), Italy (3742) and Andorra (723).
\
*...  2. The max value of TotalSlope of the five countries you identified in 1:*
France (600), Austria (304), Switzerland (412), Italy (400) and Andorra (210).

### **d) (3pts)** 

An alternative to a tree map is a sunburst plot, the principle is similar to a tree map. 

Recall from the lecture that:
- Each ring is a different level of the hierarchy
- Each segment of a ring belongs to one categorical value
- The size of a segment is either divided proportionally to a value

Now, we would like to have an overview of the attribute "TotalLifts" aggregated by different level of hierarchy.

Make a sunburst plot where
- the first hierarchy(ring) is "Country"
- the second hierarchy(ring) is "Snowparks" (whether the resort has snowparks)
- the third hierarchy(ring) is "Resort"
- the size of the segments is determined by the attribute "TotalLifts".

Then, briefly discuss your findings from the plot.

*Hint: You can use the sunburst function from plotly.express.*

In [ ]:
# your code
fig = px.sunburst(df_v, path=['Country', 'Snowparks', 'Resort'], values='TotalLifts')
fig.show()

**Your answer:**

## Question 3 - Frequent Item Sets and Association Rules (15 points)

In [ ]:
from mlxtend.preprocessing import TransactionEncoder
from mlxtend.frequent_patterns import apriori
from mlxtend.frequent_patterns import association_rules as arule

### A real online retail transaction data set of two years.

Data Set Information:
This Retail dataset contains all the transactions occurring for a UK-based and registered, non-store online retail between 01/12/2009 and 09/12/2011.The company mainly sells unique all-occasion gift ware. Many customers of the company are wholesalers.

Attribute Information:
- Invoice: Invoice number. Nominal. A 6-digit integral number is uniquely assigned to each transaction. If the number starts with 'C' it refers to a canceled transaction.
- StockCode: Product (item) code. Nominal. A 5-digit integral number uniquely assigned to each distinct product.
- Description: Product (item) name. Nominal.
- Quantity: The quantities of each product (item) per transaction. Numeric.
- InvoiceDate: Invoice date and time. Numeric. The day and time when a transaction was generated.
- Price: Unit price. Numeric. Product price per unit in sterling (£).
- CustomerID: Customer number. Nominal. A 5-digit integral number is uniquely assigned to each customer. This number has postfix 'n'.
- Country: Country name. Nominal. The name of the country where a customer resides.

### a) Loading, exploring and preprocessing the dataset (4 pts)

#### **a(i)** 
Load the data from `retail.csv` and save it under the variable `retail_df`. Display the first few lines of the dataframe.

In [ ]:
# your code

#### **a(ii)** 
To get to know the dataset, do the following:

- Show the number of rows in the dataset.
- Show the number of unique customers.
- Show the number of unique product names.
- Show the number of unique invoices.
- Show the number and the list of all the countries where the customers reside.

In [ ]:
# your code

#### **a(iii)** 
You are interested in analyzing itemsets that are frequently purchased together. Before continuing with that task, you have to make sure that the data are fit for such analysis. 1) More precisely, you want to make sure that there are no missing values in the data. 2) Moreover, you want to ensure that each item's name in the "Description" is consistent. E.g., you want "Description" values such as " coffee black", "coffee &nbsp;black", " coffee black &nbsp;", etc. to be mapped to the same value (e.g. "coffee black"). 3) Last but not least, you want to remove transactions that were canceled. Such transactions correspond to rows where the invoice number starts with letter 'C'.

Apply these preprocessing steps to the dataset `retail_df` and apply them on the dataframe itself (e.g. set inplace=True).

In [ ]:
# your code

#### **a(iv)** 
After applying the preprocessing steps in **a(iii)** , repeat again the task **a(ii)**, that is:

- Show the number of rows in the dataset.
- Show the number of unique customers.
- Show the number of all unique product names.
- Show the list of all the countries where the customers reside.

Which values changed?

In [ ]:
# your code

**Your answer:** 

### b) Frequent itemsets and Association rules (8 pts)

#### **b(i)** 
Each invoice number in the dataset identifies a unique transaction. There are potentially many rows in the dataframe having the same invoice number. We want to analyze items that are frequently purchased together, that is, items that appear in the same transaction.

Create a new dataframe named `transaction_df` with two columns: "Invoice" and "Description". Here the "Invoice" value is the index of the dataframe (the unique number identifying each row) and "Description" is the column containing all items (without duplicates) purchased within the transaction with that invoice number. Display the first few rows of your dataframe. How many rows does the `transaction_df` have?

In [ ]:
# your code

**Your answer:**

#### **b(ii)** 
Next, we want to compute frequent itemsets and association rules based on the sets of items ordered together. Use the TransactionEncoder to transform `transaction_df` into a matrix such that the value in the i-th row and the j-th column is $True$ if the i-th itemset contains product j, and $False$ otherwise. Save the matrix into a dataframe named `transactions`. Display the shape of the matrix.

*Hint: Note that your dataframe 'transactions' must contain as many rows as there are invoice numbers and as many columns as there are unique products.*

In [ ]:
# your code

#### **b(iii)** 
Use the apriori method on `transactions` to obtain all frequent itemsets using min_support=0.01. Display all frequent itemsets that have at least three items. What support count does an itemset have for our case if it satisfies min_support=0.01?

In [ ]:
# your code

**Your answer:**

#### **b(iv)**
Now we will discover association rules from the frequent itemsets. Using only the frequent itemsets with min_support=0.01 (the ones obtained in **b(iii)**), generate different association rules using min_conf=0.6 and min_conf=0.9 as thresholds. Show the association rules for each of the thresholds. What do you notice w.r.t. the number of association rules produced for the different thresholds?

In [ ]:
# your code

**Your answer:**

#### **b(v)** 
From the association rules that satisfy the confidence threshold 0.6, select and show the two rules with the highest lift. What do you notice if you compare the two rules with each other?

In [ ]:
# your code

**Your answer:**

#### **b(vi)** 

In the analysis tasks in **b)**, an itemset consisted of items that had the same invoice number (same transaction items). Thus, if an itemset was frequent, it meant that the items in it were frequently purchased together.
An association rule $A \Rightarrow B$ meant that if items in $A$ are purchased, then the items in $B$ are also purchased in that same transaction.

Suppose that we would repeat the analysis, but this time, the itemsets would consist of items having the same "CustomerID" (bought from the same customer). Interpret the meaning of the frequent itemsets and association rules for this kind of itemsets.

**Your answer:**

### c) Sequence Mining (3 pts)

For this task, the dataset used is `retail_sequences.csv`. Run the cell below to save the dataset under the dataframe `retail_sequences`. Each row in the dataframe corresponds to a unique customer (from the retail dataset). The "Customer" column contains the customer ID, whereas the "Sequence" column contains the sequence of itemsets  purchased by that customer.  Each value of "Sequence" is a sequence (list) of itemsets $<I_1, I_2, ..., I_n>$. The items within the same itemset (list without duplicates) $I_i$ were purchased together (they had the same invoice number). The itemsets are ordered by the timestamp of the transaction (value of InvoiceDate).

In [ ]:
import pandas as pd 
retail_sequences = pd.read_csv('datasets/retail_sequences.csv', converters={'Sequence': pd.eval})

Given is the sequence *s= <{'lunch bag cars blue'}, {'herb marker rosemary','herb marker thyme'}, {'wooden star christmas scandinavian'}>*. Compute the support count of that sequence, that is, compute the number of customers whose corresponding itemset sequence contains it. Display its support count and the IDs of those customers.

*Hint: In the dataset provided, all product names are unified. They are all lowercase and have no trailing spaces.*

In [ ]:
s = [['lunch bag cars blue'], ['herb marker rosemary', 'herb marker thyme'] ,['wooden star christmas scandinavian']]

In [ ]:
# your code

## Question 4: Text Mining (12 pts)

In [ ]:
from sklearn.feature_extraction.text import CountVectorizer, TfidfVectorizer
from gensim.models.doc2vec import Doc2Vec, TaggedDocument
from nltk.lm.preprocessing import padded_everygram_pipeline

### F.R.I.E.N.D.S.

In this task we will use the script from the well-known series "F.R.I.E.N.D.S.". We will apply feature extraction methods to map the line of each main character onto a vector of a vector space. Then we will train a classifier whose aim will be to predict the name of the character given a particular line from the script.
In the end, we will train language models using N-grams and produce fake sentences for each of the main characters.

### a) Data Loading and Preprocessing (4 pts)

#### **a(i)** 
Import the file `FRIENDS.csv` and save it into a dataframe named `friends_df`. Note that the dataframe must contain two columns: one indicating the character's name and one containing a line from the script. Display the first few lines from the dataframe.

<i>FYI: The script has been filtered so that it only contains lines from the main characters. The order of the lines in the data is the same as the order of the lines in the original script. Metadata and scene descriptions have been removed. Your corpus consists of all the lines contained in the data. Each row's "line" value is a single document. </i>

In [ ]:
# your code

#### **a(ii)**  
Plot the line count distribution among the six main characters (the six possible values of the column "character"). For example, show a plot containing one bar for each character whose height reflects the number of lines in `friends_df`. Briefly comment on the plot.

In [ ]:
# your code

**Your answer:**

#### **a(iii)**  
Create a corpus named `corpus` such that each document in the corpus corresponds to exactly one row's "line" in `friends_df`.

In [ ]:
# your code

#### **a(iv)**  
Write a function called `my_preprocessor` which, given a string, returns another string after tokenization, stopword removal and lemmatization have been applied. The remaining terms (tokens after stopword removal and lemmatization has been applied) should be joined in the same string using space ' '.

In [ ]:
# your code

#### **a(v)**   
You must apply your preprocessor `my_preprocessor` on each line contained in the `corpus`. Create a preprocessed corpus named `corpus_p` which contains the same lines as `corpus` after the preprocessor `my_preprocessor` has been applied.

In [ ]:
# your code

#### **a(vi)**   
Split the `friends_df` dataset from the previous task into training (80%) and test (20%) data preserving the distribution based on the "character" value. 

In [ ]:
# your code

#### **a(vii)**   
Similar to **a(iii)**, for the training data and the test data, create two corpora named `corpus_train` and `corpus_test` respectively. Each document in the training (test) corpus must correspond to exactly one row's "line" value in the corresponding training (test) dataframe.

In [ ]:
# your code

### b) Set of Words (4.5 pts)

#### **b(i)**  
We want to encode our text in such a way that for each word in the vocabulary, we are only interested in whether the word appears or not in a given document. Create such a Set of Words encoding for the whole corpus `corpus`. Use the previously defined preprocessor `my_preprocessor` as preprocessor.

In [ ]:
# your code

#### **b(ii)**  
Pick one (any) of the lines of the "line" column in the `friends_df` dataset. Display the line in:
    1) its original form, 
    2) its preprocessed version (the result contained in `corpus_p` after applying `my_preprocessor`), and 
    3) its encoding computed by the Set of Words method. This can be either an array (a vector) or a scipy matrix. 
Briefly comment on the Set of Words encoding.

In [ ]:
# your code

**Your answer:**

#### **b(iii)**  
Create a Set of Words encoding based only on the documents in `corpus_train`. Use the previously defined preprocessor `my_preprocessor` as a preprocessor.

In [ ]:
# your code

#### **b(iv)**  
In this task, we will use an SGD (Stochastic Gradient Descend) classifier to predict the character given a line from the corpus. Train the classifier on the Set of Words encoding of training corpus `corpus_train` using the character as the target feature and 'log_loss' as the loss function. Apply the classifier on the Set of Words encodings of both the training corpus and the test corpus `corpus_train`. Show its accuracy for both the training corpus and the test corpus.

In [ ]:
# your code

#### **b(v)**  
Briefly comment on the accuracy of the classifier compared to the expected accuracy of a random guesser (here: a model that simply guesses each character according to a distribution based on the line count). Use the line count distribution shown in **a(ii)** to reason about the approximate accuracy of the random guesser.

**Your answer:**

#### **b(vi)**  
Pick two lines from the dataset `friends_df`. Predict their character by applying the SGD classifier from **b(iv)** to their Set of Words encodings. Show the original lines, their original characters and the predicted characters.

In [ ]:
# your code

### c) Doc2Vec (1.5 pts)

#### **c(i)**  
In this part, we want to encode the lines using Doc2Vec. Create a Doc2Vec embedding based on the documents in the preprocessed corpus `corpus_p`. Set the vector dimension to 10 and min_count to 3.

In [ ]:
# your code

#### **c(ii)**  
Pick one (any) line from the dataset `friends_df`. Display the line and the character saying it. Find its most similar line w.r.t. the Doc2Vec encoding and display the original line and its corresponding character. Do the lines belong to the same character?

In [ ]:
# your code

**Your answer:**

### d) Language model using N-grams (2 pts)

#### For the following tasks, use the `friends_df` data.

#### **d(i)**  
For each character, create a corresponding corpus. Each corpus must be a list of documents. Each document corresponds to one "line" value of that character and it should be a list of terms. You must obtain this list of terms after applying preprocessing steps such as to lowercase, no punctuation, and tokenization to the original line. Do not perform stemming/lemmatization and/or stopword removal for this task.
Display the corpus of one of the characters.

In [ ]:
# your code

#### **d(ii)**  
For each character separately, build a trigram language model using MLE. Use both right and left padding and learn each language model using the character's corpus from **d(ii)**.

In [ ]:
# your code

#### **d(iv)**  
For each character, use the created trigram language model to generate a sentence of ten words. Display the sentences.

In [ ]:
# your code

## Question 5: Process Mining (22pts)

In [ ]:
import pandas as pd
import pm4py

In [ ]:
from pm4py.statistics.traces.generic.log import case_statistics

In [ ]:
from pm4py.algo.conformance.tokenreplay import algorithm as token_based_replay

In [ ]:
import matplotlib.pyplot as plt

### a) Loading the Data and Basic Statistics (9pts)

#### **a(i)** 
Load the data **reimburse.csv** and create a PM4Py event log. In doing so, use the following column mapping:
 - *Activity* is the activity key
 - *Case* is the case ID
 - *Timestamp* is the timestamp

In [ ]:
# your code

#### **a(ii)** 
Compute and print the following basic information:
- Number of events
- Number of cases
- Earliest timestamp
- Latest timestamp
- Number of trace variants

In [ ]:
# your code

#### **a(iii)** 
In process mining, multiple events that have the same timestamp can cause problems because the ordering of events (or even activities) becomes unclear. Moreover, they can indicate batching (i.e., one activity is executed for multiple cases simultaneously). Therefore, during your analysis, it is good to keep that in mind. To this end, compute the following statistics/answer the following questions:

1. How many events occur almost at the same time (i.e., within less than 100ms as the preceding event. (Proceeding event in the *entire* event log)?
2. Are there resources that complete two activities at the same time (within less than 100ms)?
3. How many cases are there in which two activities are executed at the same time (i.e., two events that belong to the same case occur within less than 100ms)?

*Hint: Depending on how you find the answers, be careful about event orderings.*

In [ ]:
# your code

#### **a(iv)** 
Provide a plot that shows the number of running cases (i.e., cases that have started but not yet finished) over time. You may assume that the log only contains complete traces. For each case that has started, its completion is the last observed event associated with that case. In case multiple cases start or end at the same time, you also generate multiple values.

In [ ]:
# your code

**Your answer**: *(Briefly describe the differences between the two models in about five sentences here.)*

### b) Discovery and Conformance Checking (9pts)

#### **b(i)** 
Mine a Petri net using Inductive Miner and display it.

In [ ]:
# your code

#### **b(ii)** 
Compute the fitness of the discovered Petri net using token-based replay.

*Hint: PM4Py can directly (using the top-level API) compute the number of missing, remaining, consumed, and produced tokens. Based on these, you can, for example, compute the token-based replay fitness.*

In [ ]:
# your code

#### **b(iii)** 
Filter the log to contain only traces where *Register Low* occurs. How many traces does the resulting log `log_low` contain?


In [ ]:
# your code

#### **b(iv)** 
Discover a Petri net for `log_low` and compute its fitness. How does this model differ from the model you discovered in *b(i)*? 

Suppose each of your produced process models is considered a 2-class classifier: provided a trace, it returns "Yes" if and only if the trace can be replayed by the model. Based on this perspective, how would the two process models compare in terms of precision?

In [ ]:
# your code

**Your answer:** *(Briefly describe the difference between the two models here. About two sentences can be enough.*)

**Your answer**: *(Relate your observations to precision here.)*

### Conformance Diagnostics De-jure Model
The process owner provides you a de-jure model (i.e., a model of the should-be process) and a slightly changed version of the so far considered event log. 

In this task, you will again apply conformance checking by means of token-based replay to provide diagnostics on deviations.

#### **b(v)** 
Load the Petri net *pn_conf.apnml*, the event log *log_conf.xes*, and provide the overall (i.e., model-based) token-based replay fitness score.



In [ ]:
# your code

#### **b(vi)** 
To provide additional diagnostics on the deviations, compute the missing, consumed, produced and remaining number of tokens for **each place**. 
To do so, use the following *pm4py* code:
    
    from pm4py.algo.conformance.tokenreplay import algorithm as token_based_replay
	tbr_results, place_fitness, transition_fitness, notexisting_activities_in_model =
    token_based_replay.apply(log_conf, net_conf, im_conf, fm_conf, parameters={"enable_pltr_fitness": True, "disable_variants": True})

After running this line for log `log_conf`, Petri net `net_conf` with initial marking `im_conf` and final marking `fm_conf`, the variable `place_fitness` will contain the token counts for each place and trace. Therefore, you will only need aggregate over the traces.
Print a table of the token counts per place. 


In [ ]:
# your code

#### **b(vii)** 
Consider the token counts per place and a few unfitting traces, which deviation(s) do you observe? Describe the deviation and briefly explain how it can be related to the token counts of the individual places. For example, activity *xy* is often missing resulting in a high number of missing tokens in place *p*.  


In [ ]:
# your code

**Your answer:** *(Describe the deviation(s). One sentence can already be enough.)*

**Your answer:** (*Relate the deviation(s) to the token counts of the individual places. Roughly five sentences can be enough for a precise description.*)

### c) Analyzing Fraud (4pts)

#### **c(i)** 
Filter the event log so that it only contains traces where a fraud report is filled (occurrence of `Fill Fraud Report`). For theses traces, create a bar plot showing the number of products of a certain brand involved in the fraud. Describe the resulting plot.

*Hint: Each case is associated with precisely one brand.*


In [ ]:
# your code

**Your answer:** (*Describe the plot in two to three sentences.*)

#### **c(ii)**  
The plot shows differences between brands. Discuss the result. Consider what you learned in Lecture 11 (association rules). Try to provide additional analysis results to underpin your discussion.

*Hint: A very short additional analysis (i.e., a few lines of code) might already be sufficient.*


In [ ]:
# your code (for a short additional analysis)

**Your answer:** *(Relate your results to Lecture 11, approximately one short paragraph)*

## Question 6 - Simpson's Paradox (3 pts)

### Sex Bias in Berkeley Graduate Admissions?

In the Fall of 1973, the University of California at Berkeley released data about their graduate class. The data showed the major the applicant applied to, their self-reported gender (Male or Female), and whether or not they were accepted or rejected. The acceptance rates between men and women were different. This caused immediate concern in the public as people thought that Berkeley was biased against women.

The "Berkeley Dataset" contains all 12,763 applicants to UC-Berkeley's graduate programs in Fall 1973. This dataset was published by UC-Berkeley researchers in an analysis to understand the possible gender bias in admissions.

Dataset Variables:

Year : number ➜ The application year (this value is always 1973)

Major : string ➜: An anonymized major code (either A, B, C, D, E, F, or Other). The specific majors are unknown except that A-F are the six majors with the most applicants in Fall 1973

Gender : string ➜ Applicant self-reported gender (either M or F)

Admission: string ➜ Admission decision (either Rejected or Accepted)

**a)**
Upload the data from the `berkeley.csv` file and load it into a dataframe named `data`. Display the first few lines from the dataframe.

In [ ]:
# your code

**b)** Remove the "Year" column as it does not contain any information in this dataset (all years are 1973.)

In [ ]:
# your code

**c)** For each of the values of column "Gender", compute the admission rate and compare them against each other.

In [ ]:
# your code

**Your answer:**

**d)** For each value combination of the "Gender" and "Major" columns, compute the admission rate. Compare the admission rate of women against the admission rate of men for each of the majors.

In [ ]:
# your code

**Your answer:**

**e)** Can you confirm there is a sex bias in the admission rates of the students?

**Your answer:**

## Question 7: Big Data (15pts)

In [ ]:
import pandas as pd

You are a data scientist at NASA, an agency for civil space programs, and with your team you develop and maintain the software of the NASA Crew Exploration Vehicle (CEV).  Your task is to analyze the performance of the software, and as a first exploratory step, you would like to **compute the mean execution times** of function calls within that software. Since the running vehicle will generate a high throughput of observable events in a stream, you decide to set up a MapReduce pipeline in Hadoop. 

The file **nasa-cev-software-tests.tsv** records timestamped events of the vehicle's software tests. The log contains the columns *Case*, *Activity* and *Timestamp*, denoting the case ID, the activity key (method call) and timestamp of the event record in nanoseconds, respectively. Furthermore, the log contains the columns *Lifecycle Transition* and *Execution ID*. The lifecycle transition takes either of the values *start* and *complete*, stating whether the corresponding activity (method call) in that row is being started or completed at the specified timestamp. The execution ID relates each event to a concrete method call, i.e., for each execution ID, there are exactly two entries (namely a *start* and a *complete* event) in the log.

### a) Plan the Maths (2pt):

The mean $\mu_n$ over numerical values $v_1,...,v_n$ is well-known to be computed as $\mu_n = \frac{1}{n}\sum_{i=1}^{n} v_i$.\
One may also use the alternative recursive formalization $\mu_{n+m} = \frac{n\cdot\mu_{n} + m\cdot\mu_{m}}{n+m}$. \
What is the advantage of using the alternative formalization when you think of handling streaming data or distributed data? Briefly explain.


**Your answer:**

### b) Set up MapReduce (10pts)

Realize the computation of the mean execution times of activities as a MapReduce job. 
You need to implement this a cascaded MapReduced job. This means that the output of the first job will serve as the input of the second job. In the first job, derive the execution times of each activity execution, i.e. the time difference between the *complete* and the *start* lifecycle transition of each activity execution. In the second step, aggregate this timing information to compute the mean.

#### **b(i)** 
Specify the *function signatures* of the map functions and the reduce functions that you are going to use.\
I.e., find concrete sets to substitute $K_1, V_1, ... $ in the general signatures for map and reduce functions \

$ map:  K_1 \times V_1 \rightarrow (K_2 \times V_2)^* $\
$ reduce: K_2 \times (V_2)^* \rightarrow (V_3)^*$ (or a singleton $V_3$) 

*Hint: You may introduce symbols to denote sets, e.g. $Act$ for the set of activities.\
You may also first implement the solution (b(ii)) to get an idea about the underlying signatures.\
Mind that you need two map and two reduce functions.*

**Your answer:**

#### **b(ii)**: 
Specify map functions and reduce functions to compute the mean execution time per activity as python scripts.

In [ ]:
# your code (nasa_mapper1.py)

In [ ]:
# your code (nasa_reducer1.py)

In [ ]:
# your code (nasa_mapper2.py)

In [ ]:
# your code (nasa_reducer2.py)

### c) Run MapReduce (3pts)

In the following, please use one of your team members' matriculation number as an identifier.

#### **c(i) (Randomization)**: 
Before applying your functions from the previous step to the dataset, please insert the matriculation number and run the following lines to randomly filter out a few of the traces in the event log, and continue working with the filtered log. 


In [ ]:
# your matriculation number here
matr_nb = # ... #

# utility code (do not change)
import random
random.seed(matr_nb)

full_df = pd.read_csv("datasets/nasa-cev-software-tests.tsv", sep="\t")

case_ids = list(set(full_df["Case"].values))
case_ids.sort()
filtered_out_case_ids = random.sample(case_ids, 10)
filtered_case_ids = [case_id for case_id in case_ids if case_id not in filtered_out_case_ids]
randomized_df = full_df[full_df["Case"].isin(filtered_case_ids)]

randomized_df.to_csv("datasets/nasa-cev-software-tests-randomized-" + str(matr_nb) + ".tsv",
          columns=["Activity", "Timestamp", "Lifecycle Transition", "Execution ID"],
          sep="\t", index=False, header=False)

#### **c(ii)**:
Now, it is time to work with the Hadoop Distributed File System (HDFS). Follow the instructions below and show your results in each step (screenshots of the command line).

    1) Import the event log to your Docker engine (at /usr/local/hadoop/(your_matr_nb)-event-log/). You also need to import the python scripts, but only document the event log import here.
    2) Upload the files to the running HDFS (at /input/(your_matr_nb)-event-log/).
	3) Run Hadoop commands for the MapReduce computation.
    4) Show the final output (computed mean execution times).

In [ ]:
from IPython.display import Image
# your code
# Image(filename='filename_screenshot_of_a1')

In [ ]:
# your code
# Image(filename='filename_screenshot_of_a2')

In [ ]:
# your code
# Image(filename='filename_screenshot_of_a3')

In [ ]:
# your code
# Image(filename='filename_screenshot_of_a4')